training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala

```diff
@@ -0,0 +1,220 @@
+            input : RDD[Example],
+            config : Config,
+            key : String) : FullRankLinearModel = {
+    val options = parseTrainingOptions(config.getConfig(key))
+
+    val pointwise : RDD[Example] =
+      LinearRankerUtils
+        .makePointwiseFloat(input, config, key)
+        .cache()
+
+    val model = setupModel(options, pointwise)
+
+    for (iter <- 0 until options.iterations) {
+      log.info(s"Iteration $iter")
+      modelIteration(sc, options, model, pointwise)
+    }
+    model
+  }
+
+  def trainAndSaveToFile(sc : SparkContext,
+                         input : RDD[Example],
+                         config : Config,
+                         key : String) = {
+    val model = train(sc, input, config, key)
+    TrainingUtils.saveModel(model, config, key + ".model_output")
+  }
+  
+  def modelIteration(sc : SparkContext,
+                     options : FullRankLinearTrainerOptions,
+                     model : FullRankLinearModel,
+                     pointwise : RDD[Example]) = {
+    val gradients : Array[((String, String), GradientContainer)] = options.loss match {
+      case "softmax" => softmaxGradient(sc, options, model, pointwise)
+      case _ : String => softmaxGradient(sc, options, model, pointwise)
+    }
+    val weightVector = model.getWeightVector()
+    val dim = model.getLabelDictionary.size()
+    var gradientNorm = 0.0
+    var featureCount = 0
+    // Gradient update rule from "boosting with structural sparsity Duchi et al 2009"
+    gradients.foreach(kv => {
+      val (key, gradient) = kv
+      val featureMap = weightVector.get(key._1)
+      if (featureMap != null) {
+        val weight = featureMap.get(key._2)
+        if (weight != null) {
+          // Just a proxy measure for convergence.
+          gradientNorm = gradientNorm + gradient.grad.dot(gradient.grad)
+          val scale = 2.0 / math.max(1e-6, gradient.featureSquaredSum)
+          weight.multiplyAdd(-scale.toFloat, gradient.grad)
+          val hingeScale = 1.0 - options.lambda * scale / math.sqrt(weight.dot(weight))
+          if (hingeScale <= 0.0f) {
+            // This entire weight got regularized away.
+            featureMap.remove(key._2)
+          } else {
+            featureCount = featureCount + 1
+            weight.scale(hingeScale.toFloat)
+          }
+        }
+      }
+    })
+    log.info("Sum of Gradient L2 norms = " + gradientNorm)
+    log.info("Num active features = " + featureCount)
+  }
+  
+  def softmaxGradient(sc : SparkContext,
+                      options : FullRankLinearTrainerOptions,
+                      model : FullRankLinearModel,
+                      pointwise : RDD[Example]) : Array[((String, String), GradientContainer)] = {
+    val modelBC = sc.broadcast(model)
+    
+    pointwise
+    .mapPartitions(partition => {
+      val model = modelBC.value
+      val labelToIdx = model.getLabelToIndex()
+      val dim = model.getLabelDictionary.size()
```

Comment:
Since the trainer is a singleton object I don't feel comfortable using a shared global variable in case the trainer methods are called elsewhere e.g. by an external pipeline